File: oauth2.py
from fastapi import Depends, HTTPException
from jose import JWTError, jwt
from datetime import datetime, timedelta
from fastapi.security import OAuth2PasswordBearer
from sqlalchemy.orm import session
from dotenv import load_dotenv
import os

import sys
sys.path.append('BackEnd')

import models
import DataBase
import oauth2
import utils
import schemas
from routes import user
load_dotenv()
# Define your SECRET_KEY and ALGORITHM
SECRET_KEY = os.getenv('SECRET_KEY')
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 50000

# Create an access token
def create_access_token(data: dict):
    to_encode = data.copy()
    expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({'exp': expire})

    encoded_token = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_token

# Verify an access token
def verify_access_token(id: int, token: str, credentials_exception=None):
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        user_id: int = payload.get("user_id")

        if not user_id:
            if credentials_exception:
                raise credentials_exception
            else:
                return False
        if user_id != id:
            return False
        
        return True
    
    except JWTError:
        if credentials_exception:
            raise credentials_exception
        else:
            return False
        

File: Backend(Express)/oauth2.js
const jwt = require('jsonwebtoken');
const dotenv = require('dotenv');
const { DateTime } = require('luxon');

dotenv.config();
const SECRET_KEY = process.env.SECRET_KEY;
const ALGORITHM = 'HS256';
const ACCESS_TOKEN_EXPIRE_MINUTES = 50000;

// Create an access token
function createAccessToken(data) {
    if (typeof data !== 'object' || data === null) {
        throw new Error('Data must be an object');
    }

    const { user_id } = data; // Check for user_id
    if (user_id === undefined || user_id === null) {
        throw new Error('user_id is missing from the data');
    }

    const toEncode = { ...data };
    const expire = DateTime.utc().plus({ minutes: ACCESS_TOKEN_EXPIRE_MINUTES }).toJSDate();
    toEncode.exp = expire.getTime() / 1000;

    const encodedToken = jwt.sign(toEncode, SECRET_KEY, { algorithm: ALGORITHM });
    return encodedToken;
}




// Verify an access token
function verifyAccessToken(token, credentialsException = null) {

    try {
        const payload = jwt.verify(token, id, SECRET_KEY, { algorithms: [ALGORITHM] });
        const userId = payload.user_id;
        if (!userId || userId !== id) {
            if (credentialsException) {
                throw credentialsException;
            } else {
                return false;
            }
        }

        return true;
    } catch (err) {
        console.error('Error in token verification:', err); // Log any errors
        if (credentialsException) {
            throw credentialsException;
        } else {
            return false;
        }
    }
}


const authenticateToken = (req, res, next) => {
    const authHeader = req.headers['authorization'];
    if (!authHeader) return res.status(401).json({ detail: "Authorization header missing" });

    const token = authHeader.split(" ")[1];
    if (!token) return res.status(401).json({ detail: "Token not provided" });

    // Attach the token to the request for later use
    req.token = token;
    next();
};

module.exports = { createAccessToken, verifyAccessToken, authenticateToken };
